Add state spaces Sphere, Torus and Mobius.

aorthey opened the pull request:
Hello everyone,

to make sure that a planner works correctly, I found it super useful to have some simple geometrical spaces which we can visualize. Here are three which I often use: Sphere, Torus and Mobius strip. They all come with correct uniform sampling methods and correct interpolation/distance metrics attached (see pictures to get an idea). 

I thought someone else might be interested in using them, so I wanted to share them. 

Best,
Andreas

![sphere_interpolation_thumbnail](https://user-images.githubusercontent.com/1220541/109400707-7573db80-794a-11eb-85fa-872cf6382958.png)![sphere_samples_thumbnail](https://user-images.githubusercontent.com/1220541/109400708-7573db80-794a-11eb-991a-3319c1eeaa4c.png)![torus_interpolation_thumbnail](https://user-images.githubusercontent.com/1220541/109400710-760c7200-794a-11eb-9972-89bc57dfc538.png)![torus_samples_thumbnail](https://user-images.githubusercontent.com/1220541/109400712-760c7200-794a-11eb-8f17-1f3859bda15f.png)![mobius_interpolation_thumbnail](https://user-images.githubusercontent.com/1220541/109400705-7442ae80-794a-11eb-8105-12d544de441b.png)![mobius_samples_thumbnail](https://user-images.githubusercontent.com/1220541/109400706-7573db80-794a-11eb-94e7-859c8c3cc6fc.png)

mamoll commented on src/ompl/base/spaces/src/MobiusStateSpace.cpp:
```diff
@@ -0,0 +1,146 @@
+    return state;
+}
+
+void MobiusStateSpace::freeState(State *state) const
```
Do you need this method? The method doesn't override anything the base class would do.

mamoll commented on src/ompl/base/spaces/src/SphereStateSpace.cpp:
```diff
@@ -0,0 +1,132 @@
+    return state;
+}
+
+void ompl::base::SphereStateSpace::freeState(State *state) const
```
Not needed

mamoll commented on src/ompl/base/spaces/src/TorusStateSpace.cpp:
```diff
@@ -0,0 +1,142 @@
+    // https://stackoverflow.com/questions/26300510/generating-random-points-on-a-surface-of-an-n-dimensional-torus
+    // Based on publication "Random selection of points distributed on curved surfaces."
+    // Link: https://iopscience.iop.org/article/10.1088/0031-9155/32/10/009/pdf
+    const TorusStateSpace *T = static_cast<const TorusStateSpace *>(space_);
```
```suggestion
    const auto *T = static_cast<const TorusStateSpace *>(space_);
```

mamoll commented on src/ompl/base/spaces/src/TorusStateSpace.cpp:
```diff
@@ -0,0 +1,142 @@
+    return state;
+}
+
+void TorusStateSpace::freeState(State *state) const
```
Not needed

mamoll reviewed (changes requested):
Looks good! I have left some minor comments. I hope to get to your older massive PR soon.

aorthey commented:
Thank you Mark! Not sure why freeState was still in there. Should be good now.

mamoll commented:
If it's not too much trouble, could you extend the [state space tests](https://github.com/ompl/ompl/blob/main/tests/base/state_spaces.cpp) with some basic tests for your new spaces?

marlinstrub commented:
Thanks for sharing these @aorthey

aorthey commented:
Ok I added the state space tests. I found two small bugs (Mobius used same header define as Torus, and the sphere distance function used a wrong angle). Now it looks good and tests run through.

I also added a function "toVector" which projects the parameterization of the spaces into 3D space (which would be useful if anyone wants to actually visualize it).
